src/Controller/EntityCreateController.php

```diff
@@ -0,0 +1,170 @@
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Entity\EntityTypeBundleInfoInterface;
+use Drupal\Core\Render\RendererInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Url;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * A generic controller for creating entities.
+ */
+class EntityCreateController extends ControllerBase {
+
+  /**
+   * The entity type bundle info.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface
+   */
+  protected $entityTypeBundleInfo;
+
+  /**
+   * The renderer service.
+   *
+   * @var \Drupal\Core\Render\RendererInterface
+   */
+  protected $renderer;
+
+  /**
+   * Constructs a new EntityCreateController object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $entity_type_bundle_info
+   *   The entity type bundle info.
+   * @param \Drupal\Core\Render\RendererInterface $renderer
+   *   The renderer.
+   */
+  public function __construct(EntityTypeBundleInfoInterface $entity_type_bundle_info, RendererInterface $renderer) {
+    $this->entityTypeBundleInfo = $entity_type_bundle_info;
+    $this->renderer = $renderer;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.bundle.info'),
+      $container->get('renderer')
+    );
+  }
+
+  /**
+   * Displays add links for the available bundles.
+   *
+   * Redirects to the add form if there's only one bundle available.
+   *
+   * @param string $entity_type_id
+   *   The entity type ID.
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request.
+   *
+   * @return \Symfony\Component\HttpFoundation\RedirectResponse|array
+   *   If there's only one available bundle, a redirect response.
+   *   Otherwise, a render array with the add links for each bundle.
+   */
+  public function addPage($entity_type_id, Request $request) {
+    $entity_type = $this->entityTypeManager()->getDefinition($entity_type_id);
+    $bundle_type = $entity_type->getBundleEntityType();
+    $form_route_name = 'entity.' . $entity_type_id . '.add_form';
+    $build = [
+      '#theme' => 'entity_add_list',
+      '#cache' => [
+        'tags' => $entity_type->getListCacheTags(),
+      ],
+      '#bundle_type' => $bundle_type,
+      '#form_route_name' => $form_route_name,
+    ];
+    $bundles = array_keys($this->entityTypeBundleInfo->getBundleInfo($entity_type_id));
+    // Filter out the bundles the user doesn't have access to.
+    $access_control_handler = $this->entityTypeManager()->getAccessControlHandler($bundle_type);
+    foreach ($bundles as $index => $bundle_name) {
+      $access = $access_control_handler->createAccess($bundle_name, NULL, [], TRUE);
+      if (!$access->isAllowed()) {
+        unset($bundles[$index]);
+      }
+      $this->renderer->addCacheableDependency($build, $access);
```

Comment:
Do you think we should move this entire segment from preprocess to the controller:
```php
  foreach ($variables['bundles'] as $bundle) {
    $variables['bundles'][$bundle->id()] = [
      'bundle' => $bundle->id(),
      'add_link' => Link::createFromRoute($bundle->label(), $variables['form_route_name'], [$bundle_type => $bundle->id()]),
      'description' => [
        '#markup' => $bundle->getDescription(),
      ],
    ];
  }
```
Felt cleaner to do it the way it is right now.